python.md

```diff
@@ -0,0 +1,122 @@
+---
+layout: default
+title: rpm.org - Python RPM
+---
+
+# Python RPM
+
+A showcase of the Python RPM on real-life examples.
+
+
+## Querying a single package
+
+Query a single installed package by its name. A Python equivalent for
+`rpm -q hello`:
+
+```
+import rpm
+ts = rpm.TransactionSet()
+mi = ts.dbMatch("name", "hello")
+hdr = next(mi, None)
+if hdr:
+    print(hdr[rpm.RPMTAG_NVRA])
+```
+
+
+## Querying a single package
+
+Again, query a single package but this time print more information. A
+Python equivalent for `rpm -qi hello`:
+
+
+```python
+import rpm
+ts = rpm.TransactionSet()
+mi = ts.dbMatch("name", "hello")
+hdr = next(mi, None)
+if hdr:
+    print(
+        f"Name        : {hdr[rpm.RPMTAG_NAME]}\n"
+        f"Version     : {hdr[rpm.RPMTAG_VERSION]}\n"
+        f"Release     : {hdr[rpm.RPMTAG_RELEASE]}\n"
+        f"Architecture: {hdr[rpm.RPMTAG_ARCH]}\n"
+        f"Install Date: {hdr[rpm.RPMTAG_INSTALLTIME]}\n"
+        f"Group       : {hdr[rpm.RPMTAG_GROUP]}\n"
+        f"Size        : {hdr[rpm.RPMTAG_SIZE]}\n"
+        f"License     : {hdr[rpm.RPMTAG_LICENSE]}\n"
+        f"Signature   : TODO\n"
+        f"Source RPM  : {hdr[rpm.RPMTAG_SOURCERPM]}\n"
+        f"Build Date  : {hdr[rpm.RPMTAG_BUILDTIME]}\n"
+        f"Build Host  : {hdr[rpm.RPMTAG_BUILDHOST]}\n"
+        f"Packager    : {hdr[rpm.RPMTAG_PACKAGER]}\n"
+        f"Vendor      : {hdr[rpm.RPMTAG_VENDOR]}\n"
+        f"URL         : {hdr[rpm.RPMTAG_URL]}\n"
+        f"Bug URL     : {hdr[rpm.RPMTAG_BUGURL]}\n"
+        f"Summary     : {hdr[rpm.RPMTAG_SUMMARY]}\n"
+        f"Description :\n{hdr[rpm.RPMTAG_DESCRIPTION]}\n"
+    )
```

Comment:
How challenging would it be to provide an example of iterating over files in the payload, or changelog entries?

Reply:
So basically examples showing how to implement `rpm -ql hello` and `rpm -q --changelog hello` in Python? No idea, I haven't needed it so far. But I will try to add them as well.

Reply:
There's no high-level API for changelogs, but iterating over them is just walking through the three header arrays the data is stored in. This dumps the raw data, for real-world example you'd probably want to pretty it up a bit:

```
chl = zip(h['changelogtime'], h['changelogname'], h['changelogtext'])
for item in chl:
    print(item)
```